### PR TITLE
Remodelling Observers/Examples that rely on UB

### PR DESCRIPTION
**Pull request by domenukk**
This tries to Address #949 and resumes on #446

### CODE REVIEW

File: libafl_targets/src/coverage.rs [hunk 1]
```diff
@@ -53,18 +55,43 @@ pub fn autotokens() -> Result<Tokens, Error> {
...
-    OwnedSliceMut::from_raw_parts_mut(EDGES_MAP_PTR, EDGES_MAP_PTR_SIZE)
+pub unsafe fn edges_map_mut_slice<'a>() -> OwnedMutSlice<'a, u8> {
+    OwnedMutSlice::from_raw_parts_mut(edges_map_mut_ptr(), edges_max_num())
+}
+
+/// Gets a new [`StdMapObserver`] from the current [`edges_map_mut_slice`].
+///
+/// # Safety
+/// This will dereference [`edges_map_mut_ptr`] and crash if it is not a valid address.
+pub unsafe fn std_edges_map_observer<'a, S>(name: S) -> StdMapObserver<'a, u8, false>
```
Comment (domenukk) [inline libafl_targets/src/coverage.rs:76]: @andreafioraldi can you double check if this whole looks correct? Thx

File: libafl/src/observers/map.rs [hunk 2]
```diff
@@ -444,14 +444,29 @@ where
...
-    fn maybe_differential<S>(name: S, map: &'a mut [T]) -> Self
+    unsafe fn maybe_differential<S>(name: S, map: &'a mut [T]) -> Self
     where
         S: Into<String>,
     {
-        Self {
-            map: OwnedSliceMut::from(map),
+        let len = map.len();
+        let ptr = map.as_mut_ptr();
+        Self::maybe_differential_from_mut_ptr(name, ptr, len)
```
Comment (domenukk) [inline libafl/src/observers/map.rs:458]: @VTCAKAVSMoACE is something like this sound?

File: libafl/src/observers/map.rs [hunk 3]
```diff
@@ -444,14 +444,29 @@ where
...
+    }
+
+    /// Creates a new [`MapObserver`] from an [`OwnedMutSlice`]
+    #[must_use]
+    fn maybe_differential_from_mut_slice<S>(name: S, map: OwnedMutSlice<'a, T>) -> Self
```
Comment (domenukk) [inline libafl/src/observers/map.rs:463]: These APIs are new and should be sound, as long as they are called with a `OwnedMutSlice::RefRaw`

Really, we probably want a way to `Rc<RefCell<>>` these, too(?)

File: libafl_targets/src/coverage.rs (hunk 1 above)
Comment (domenukk) [reply libafl_targets/src/coverage.rs:76]: Also, should this use `ConstMapObserver` on non-ptr maps?

File: libafl/src/observers/mod.rs [hunk 4]
```diff
@@ -475,25 +475,30 @@ where
...
     T: Debug + Serialize + serde::de::DeserializeOwned,
 {
     /// Creates a new [`ListObserver`] with the given name.
+    ///
+    /// # Safety
+    /// Will dereference the list.
+    /// The list may not move in memory.
     #[must_use]
-    pub fn new(name: &'static str, list: &'a mut Vec<T>) -> Self {
+    pub unsafe fn new(name: &'static str, list: *mut Vec<T>) -> Self {
```
Comment (domenukk) [inline libafl/src/observers/mod.rs:498]: @VTCAKAVSMoACE can we somehow enforce the `Vec` to be `Pin` and live for `'a`?

**Comment by domenukk** [comment]
We may want to enforce `Pin` in the APIs?

File: libafl_targets/src/coverage.rs (hunk 1 above)
Comment (andreafioraldi) [reply libafl_targets/src/coverage.rs:76]: > Also, should this use ConstMapObserver on non-ptr maps?

No, slices are not const, const maps is when the size is set at compile time.

> @andreafioraldi can you double check if this whole looks correct? Thx

tomorrow, I took vacation today

**Comment by andreafioraldi** [comment]
As we are doing this, we should consider if it we should pass the observers tuple to the harness function in inprocess executor. in that case we can store a mut ref in the observer as it is modified in the harness without UB.

**Comment by domenukk** [comment]
> As we are doing this, we should consider if it we should pass the observers tuple to the harness function in inprocess executor. in that case we can store a mut ref in the observer as it is modified in the harness without UB.

I thought you already had a PR for that and decided against it, as you can access them from a global?

Technically, having observers that are `Rc<RefCell<..>>` can be another easy solution for shared access

**Comment by domenukk** [comment]
@andreafioraldi comments?

**Comment by andreafioraldi** [comment]
> @andreafioraldi comments?

lgtm

File: .github/workflows/build_and_test.yml [hunk 5]
```diff
@@ -174,8 +174,9 @@ jobs:
     - name: Add no_std toolchain
       run: rustup toolchain install nightly-x86_64-unknown-linux-gnu ; rustup component add rust-src --toolchain nightly-x86_64-unknown-linux-gnu
     - name: Install python
+      # Removing macOS things already installed in CI against failed linking
       if: runner.os == 'macOS'
-      run: brew install --force-bottle --overwrite python@3.11
+      run: rm /usr/local/bin/2to3 /usr/local/bin/idle3 /usr/local/bin/pydoc3 /usr/local/bin/python3 /usr/local/bin/python3-config; brew install --force-bottle --overwrite python
```
Comment (domenukk) [inline .github/workflows/build_and_test.yml:179]: Cc @tokatoka

**Pull request merged**